### PR TITLE
Add option for bridging all topics

### PR DESCRIPTION
**Pull request by dhood**
without this, getting a topic to be bridged requires starting up temporary subscribers on the other side. introspection tools e.g. `rostopic list` can't see ROS 2 publishers, and tools such as `rosbag record --all` aren't very usable.

With this, bridges will be made for any topic that has a known mapping, whether subscribers are on the other side or not.

right now it is trying to make a bridge for a topic even if it doesn't have a known mapping, so you get e.g. `failed to create 1to2 bridge for topic 'rosout_agg' with ROS 1 type 'rosgraph_msgs/Log' and ROS 2 type '': No template specialization for the pair` printed out. I will fix it by adding a function that check if a factory exists for the type before forcibly creating the mapping.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
This approach could become problematic if we ever have something beyond a 1:1 mapping. If a single type in one ROS version can be mapped to more than one type in the other version it is unknown which mapping should be created.

**Comment by dhood** [comment]
Thanks for the point about the non-1:1 mappings. it doesn't seem like it will be much effort to make [this new method](https://github.com/ros2/ros1_bridge/pull/42/commits/8485326b578cc65c77380f75cf59f7fbaa05ed17#diff-ee525d02c311b44f9fefa22ad669d365R50) return all possible mappings and forcing them all - what do you think?

Also, what do you think is best for this signature https://github.com/ros2/ros1_bridge/pull/42/commits/8485326b578cc65c77380f75cf59f7fbaa05ed17#diff-ee525d02c311b44f9fefa22ad669d365R56 : how it is now, where outputs are listed after inputs, or switching them so that `ros1_type_name` comes before `ros2_type_name` (for consistency with `get_1to2_mapping`)

**Comment by dhood** [comment]
alternatively we could not support non-1:1 mappings at this point. I think we will already run into issues on master if you have multiple mappings from ROS 2 to ROS 1 because the ROS 1 subscribers all get assigned type `""`

**Comment by dirk-thomas** [comment]
> Thanks for the point about the non-1:1 mappings. it doesn't seem like it will be much effort to make [this new method](https://github.com/ros2/ros1_bridge/pull/42/commits/8485326b578cc65c77380f75cf59f7fbaa05ed17#diff-ee525d02c311b44f9fefa22ad669d365R50) return all possible mappings and forcing them all - what do you think?

That sounds good.

> Also, what do you think is best for this signature https://github.com/ros2/ros1_bridge/pull/42/commits/8485326b578cc65c77380f75cf59f7fbaa05ed17#diff-ee525d02c311b44f9fefa22ad669d365R56 : how it is now, where outputs are listed after inputs, or switching them so that `ros1_type_name` comes before `ros2_type_name` (for consistency with `get_1to2_mapping`)

For a function calles `get_XtoY_mapping` I would expect the arguments to be in the order as mentioned in the name, so first `X` and then `Y`. The fact that there is also a function `getYtoX_mapping` doesn't matter imo.

**Comment by dhood** [comment]
I've rebased this after https://github.com/ros2/ros1_bridge/pull/44
I haven't accommodated non-1:1 mappings but think that this could still be reviewed as-is

File: src/dynamic_bridge.cpp [hunk 1]
```diff
@@ -51,6 +51,50 @@ struct Bridge2to1HandlesAndMessageTypes
...
+  return std::find(args.begin(), args.end(), option) != args.end();
+}
+
+bool get_flag_option(const std::vector<std::string> & args, const std::string & option)
+{
+  auto it = std::find(args.begin(), args.end(), option);
+  if (it != args.end()) {
+    return true;
+  }
+  return false;
```
Comment (dirk-thomas) [inline src/dynamic_bridge.cpp]: These lines can be written as:

```return it != args.end()```

File: src/dynamic_bridge.cpp [hunk 2]
```diff
@@ -60,23 +104,36 @@ void update_bridge(
...
     // identify topics available as ROS 1 publishers as well as ROS 2 subscribers
     auto topic_name = ros1_publisher.first;
+    std::string ros1_type_name = ros1_publisher.second;
+    std::string ros2_type_name;
+
     auto ros2_subscriber = ros2_subscribers.find(topic_name);
     if (ros2_subscriber == ros2_subscribers.end()) {
-      continue;
+      if (bridge_all_1to2_topics) {
+        // update the ROS 2 type name to be the anticipate the bridged type
```
Comment (dirk-thomas) [inline src/dynamic_bridge.cpp]: Missing word?

File: src/dynamic_bridge.cpp [hunk 3]
```diff
@@ -60,23 +104,36 @@ void update_bridge(
...
+        bool mapping_found = ros1_bridge::get_1to2_mapping(ros1_type_name, ros2_type_name);
+        if (!mapping_found) {
+          // printf("No known mapping for ROS 1 type '%s'\n", ros1_type_name.c_str());
+          continue;
+        }
+        // printf("topic name '%s' has ROS 2 publishers\n", topic_name.c_str());
+      } else {
+        continue;
+      }
```
Comment (dirk-thomas) [inline src/dynamic_bridge.cpp]: The nesting is pretty deep with late `else` blocks and `continue`. I would suggest inverting the logic to `continue` early:

```
if (!bridge_all_1to2_topics) {
  continue;
}
bool mapping_found = ros1_bridge::get_1to2_mapping(ros1_type_name, ros2_type_name);
if (!mapping_found) {
  continue;
}
```

File: src/dynamic_bridge.cpp [hunk 4]
```diff
@@ -114,18 +171,31 @@ void update_bridge(
...
+    std::string ros2_type_name = ros2_publisher.second;
+    std::string ros1_type_name;
+
+    auto ros1_subscriber = ros1_subscribers.find(topic_name);
+    if (ros1_subscriber == ros1_subscribers.end()) {
+      if (bridge_all_2to1_topics) {
+        // update the ROS 1 type name to be the anticipate the bridged type
+        bool mapping_found = ros1_bridge::get_2to1_mapping(ros2_type_name, ros1_type_name);
+        if (!mapping_found)
+        {
```
Comment (dirk-thomas) [inline src/dynamic_bridge.cpp]: Same here. And the `{` should be trailing the previous line.

File: src/dynamic_bridge.cpp [hunk 5]
```diff
@@ -200,6 +270,15 @@ void update_bridge(
 
 int main(int argc, char * argv[])
 {
+  bool output_topic_introspection;
+  bool bridge_all_1to2_topics;
+  bool bridge_all_2to1_topics;
+  if (!parse_command_options(argc, argv,
+    output_topic_introspection, bridge_all_1to2_topics, bridge_all_2to1_topics))
```
Comment (dirk-thomas) [inline src/dynamic_bridge.cpp]: Please wrap before `argc` ("either all or none but not some").

File: ros1_bridge/__init__.py [hunk 6]
```diff
@@ -114,6 +114,10 @@ def generate_cpp(output_path, template_dir):
     template_file = os.path.join(template_dir, 'get_factory.cpp.em')
     expand_template(template_file, data, os.path.join(output_path, 'get_factory.cpp'))
 
+    data = {'mappings': ordered_mappings}
+    template_file = os.path.join(template_dir, 'get_mappings.cpp.em')
```
Comment (dirk-thomas) [inline ros1_bridge/__init__.py]: Where is this template?

Comment (dhood) [inline ros1_bridge/__init__.py]: oops - added in https://github.com/ros2/ros1_bridge/pull/42/commits/e0c68e870878896533b34c7e5396a24a2f1f1d5e

**Comment by dhood** [comment]
Merge conflicts addressed after services were added.
I am not sure why https://github.com/ros2/ros1_bridge/pull/42/commits/4d26a312210a3e381806f08e03dc75e81f380867 was necessary to avoid empy `em.Error: interpreter globals collision` errors but I can confirm that bridging all topics is working for the packaged bridge.

Here's a job with the `bridge-all*` options: http://ci.ros2.org/view/packaging/job/packaging_linux-ci/1/

**Comment by dirk-thomas** [comment]
You can't reuse the same dictionary for multiple empy evaluations. You should pass a "new" dict to each evaluate call instead.

**Comment by dhood** [comment]
ah, thank you :smile:
in that case I propose we merge this.
When testing this I did notice that the service bridging wasn't working (at least now how I expected it to). I ran a job from master and confirmed it wasn't only on this branch. (I mention it in case you're testing the packaging job and wondering the same)

**Comment by dirk-thomas** [comment]
Sounds good to me.

**Pull request merged**